File: src/main/java/pl/laskowski/marcin/parsing/text/SudokuTextFormatParser.java
package pl.laskowski.marcin.parsing.text;

import pl.laskowski.marcin.model.Sudoku;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Pattern;

/**
 * Created by Marcin Laskowski.
 */

public class SudokuTextFormatParser {

    private static final Pattern EMPTY_VALUE = Pattern.compile("^_*$");
    private static final Pattern NO_FIELD = Pattern.compile("^#*$");

    public Sudoku parseOne(String text) {
        return new Command(text).parseOne();
    }

    public List<Sudoku> parseMany(String text) {
        return new Command(text).parseMany();
    }

    public Sudoku parseOne(File file) throws FileNotFoundException {
        return new Command(file).parseOne();
    }

    public List<Sudoku> parseMany(File file) throws FileNotFoundException {
        return new Command(file).parseMany();
    }

    private class Command {

        private final Scanner scanner;
        private Scanner lineScanner;

        private int width;
        private int height;
        private Integer[] data;
        private int pointer;

        private Command(String text) {
            this.scanner = new Scanner(text);
        }

        private Command(File file) throws FileNotFoundException {
            this.scanner = new Scanner(file);
        }

        private List<Sudoku> parseMany() {
            List<Sudoku> sudokuList = new ArrayList<>();
            while (scanner.hasNext()) {
                Sudoku sudoku = parseOne();
                sudokuList.add(sudoku);
            }
            return sudokuList;
        }

        private Sudoku parseOne() {
            readSize();
            for (int y = 0; y < height; y++) {
                String line = scanner.nextLine();
                if (line.isEmpty()) {
                    y--;
                    continue;
                }
                lineScanner = new Scanner(line).useDelimiter(",");
                for (int x = 0; x < width; x++, pointer++) {
                    Integer value = parse(lineScanner.next().trim());
                    data[pointer] = value;
                }
            }
            return new Sudoku(width, height, data);
        }

        private void readSize() {
            width = scanner.nextInt();
            height = scanner.nextInt();
            data = new Integer[width * height];
            pointer = 0;
            scanner.nextLine();
        }

        private Integer parse(String input) {
            if (isEmptyValue(input)) {
                return 0;
            } else if (isNoField(input)) {
                return null;
            } else {
                return asNumber(input);
            }
        }

        private boolean isEmptyValue(String input) {
            return EMPTY_VALUE.matcher(input).matches();
        }

        private boolean isNoField(String input) {
            return NO_FIELD.matcher(input).matches();
        }

        private int asNumber(String input) {
            return Integer.parseInt(input, 10);
        }

    }

}

File: app/src/main/java/pl/laskowski/marcin/parsing/text/SudokuTextFormatParser.kt
package pl.laskowski.marcin.parsing.text

import pl.laskowski.marcin.model.Sudoku
import pl.laskowski.marcin.type.SudokuTypes
import pl.laskowski.marcin.type.SudokuVariant
import java.io.File
import java.io.FileNotFoundException
import java.util.*
import java.util.regex.Pattern

class SudokuTextFormatParser {

    fun parseOne(text: String): Sudoku {
        return Command(text).parseOne()
    }

    fun parseMany(text: String): List<Sudoku> {
        return Command(text).parseMany()
    }

    @Throws(FileNotFoundException::class)
    fun parseOne(file: File): Sudoku {
        return Command(file).parseOne()
    }

    @Throws(FileNotFoundException::class)
    fun parseMany(file: File): List<Sudoku> {
        return Command(file).parseMany()
    }

    private fun getType(type: String): SudokuVariant {
        return when(type) {
            "classic_1x1" -> SudokuTypes.CLASSIC_1x1
            "classic_2x2" -> SudokuTypes.CLASSIC_2x2
            "classic_4x4" -> SudokuTypes.CLASSIC_4x4
            "classic_9x9" -> SudokuTypes.CLASSIC_9x9
            else -> throw IllegalArgumentException("Unknown sudoku type: $type")
        }
    }

    private inner class Command {
        private val scanner: Scanner
        private lateinit var type: SudokuVariant
        private var width = 0
        private var height = 0
        private lateinit var data: MutableList<Int?>
        private var pointer = 0

        constructor(text: String) {
            scanner = Scanner(text)
        }

        constructor(file: File) {
            scanner = Scanner(file)
        }

        fun parseMany(): List<Sudoku> {
            val sudokuList: MutableList<Sudoku> = ArrayList()
            while (scanner.hasNext()) {
                val sudoku = parseOne()
                sudokuList.add(sudoku)
                if (scanner.hasNext()) scanner.nextLine()
            }
            return sudokuList
        }

        fun parseOne(): Sudoku {
            readSize()
            var y = 0
            while (y < height) {
                val line = scanner.nextLine()
                if (line.isEmpty()) continue
                val lineScanner = Scanner(line).useDelimiter(",")
                var x = 0
                while (x < width) {
                    val value = parse(lineScanner.next().trim())
                    data[pointer] = value
                    x++
                    pointer++
                }
                y++
            }
            return Sudoku(type, data)
        }

        private fun readSize() {
            type = getType(scanner.nextLine())
            width = type.sizeX
            height = type.sizeY
            data = MutableList(width * height) { null }
            pointer = 0
        }

        private fun parse(input: String): Int? {
            return when {
                isEmptyValue(input) -> 0
                isNoField(input) -> null
                else -> asNumber(input)
            }
        }

        private fun isEmptyValue(input: String): Boolean {
            return EMPTY_VALUE.matcher(input).matches()
        }

        private fun isNoField(input: String): Boolean {
            return NO_FIELD.matcher(input).matches()
        }

        private fun asNumber(input: String): Int {
            return input.toInt(10)
        }
    }

    companion object {
        private val EMPTY_VALUE = Pattern.compile("^_*$")
        private val NO_FIELD = Pattern.compile("^#*$")
    }
}